### PR TITLE
node: ignore "deprecated" IPv6 IPs when determining primary IP

### PR DESCRIPTION
**Pull request by danwinship**
Bare metal OCP adds temporary IPs to the primary interface for use with external load-balancing. These should be ignored when determining the primary node IP.

I was going to fix this by implementing #1690 but there's currently a messy tangle of different ways IPs are detected (`util.GetNodePrimaryIP`, `getDefaultIfAddr`, and `getNetworkInterfaceIPAddresses` each do something different and are each only used in one place) and it needs more rethinking than I wanted to do right now, so I just fixed this one method to be better.

### CODE REVIEW

**Comment by dcbw** [comment]
@danwinship do we expect to fail those ~6 conformance tests in v6?  if so, just merge this.

**Comment by aojea** [comment]
we are not testing local ipv6 :thinking:

File: go-controller/pkg/util/net_linux.go [hunk 1]
```diff
@@ -318,3 +319,28 @@ func DeleteConntrack(ip string, port int32, protocol kapi.Protocol) error {
...
+	}
+
+	addrs, err := netLinkOps.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list addresses for %q: %v", iface, err)
+	}
+
+	var ips []*net.IPNet
+	for _, addr := range addrs {
+		if addr.Scope != unix.RT_SCOPE_UNIVERSE {
```
Comment (aojea) [inline go-controller/pkg/util/net_linux.go:337]: this does not rule out ULA, right?

**Comment by aojea** [comment]
https://github.com/ovn-org/ovn-kubernetes/pull/1719

File: go-controller/pkg/util/net_linux.go [hunk 2]
```diff
@@ -318,3 +319,28 @@ func DeleteConntrack(ip string, port int32, protocol kapi.Protocol) error {
...
+			continue
+		}
+		if (addr.Flags & unix.IFA_F_DEPRECATED) != 0 {
```
Comment (trozet) [inline go-controller/pkg/util/net_linux.go:340]: maybe a comment on this as to why its here would be helpful in the future. Do we actually have "deprecated" addresses on interfaces in downstream. I've never seen that before.

Comment (danwinship) [reply go-controller/pkg/util/net_linux.go]: On bare metal jobs where there is no cloud load balancer, it is common to add additional IPs to the node to accept load balancer connections, and to sometimes rebalance those LB IPs between nodes (especially if a node goes down), using ARPs or BGP or whatever to track their movement between nodes.

In the IPv4 case they use `IFA_F_SECONDARY` on these IPs, which seems to automatically cause the IP to move to the end of the list of the interface's IPs, so any code that just picks the first IP will never see those IPs anyway so it was never a problem (though, in the latest version of this patch I added a check for that too, just for completeness).

In IPv6 they use `IFA_F_DEPRECATED` and not `IFA_F_SECONDARY`. (It's possible "secondary" doesn't exist for IPv6 addresses?) "Deprecated" was originally intended for use when you're rolling over IP addresses; you add a new IP, mark the old one as "deprecated", and then when all the connections to the old IP have been closed, and you don't expect any more to come in, you remove it. Among other effects, marking an address as "deprecated" means the kernel will never use it as the source IP of a new outbound connection. So, these semantics are not quite right for a transient load-balancer IP, but they're pretty good; it signals an intention that the IP isn't expected to be around forever, and it keeps the kernel from using that IP for things it wasn't meant for. And the "not expected to be around forever" means we shouldn't use it as the primary node IP either, so it's right to check that flag here.

(Added a comment mentioning bare metal load balancing... people can `git blame` and find the PR for the longer story...)

File: go-controller/pkg/util/net_linux.go (hunk 1 above)
Comment (danwinship) [reply go-controller/pkg/util/net_linux.go]: > this does not rule out ULA, right?

It's not about the address range, it's about the flags that were used when adding it... I rewrote it to check `IsLinkLocalUnicast` instead.

**Comment by danwinship** [comment]
(tests passed)

**Review by trozet — APPROVED** [review]
lgtm

**Pull request merged**